### PR TITLE
Dir up down

### PR DESCRIPTION
**Pull request by Zstro**

### CODE REVIEW

File: src/S3ClientGUI/mainwindow.cpp [hunk 1]
```diff
@@ -333,6 +333,7 @@ void MainWindow::on_S3ContextMenuRequest(const QPoint &point)
     } else if (item->type == S3ParentDirectoryType)
         menu.addAction("Open", this, SLOT(on_open()));
     else if (item->type == S3DirectoryType) {
+        menu.addAction("Download", this, SLOT(on_downloadDir()));
```
Comment (thesues) [inline src/S3ClientGUI/mainwindow.cpp:336]: OnDownloadDir is better

File: src/S3ClientGUI/mainwindow.cpp [hunk 2]
```diff
@@ -356,12 +357,83 @@ void MainWindow::on_LocalContextMenuRequest(const QPoint &point)
...
         menu.exec(m_fsview->mapToGlobal(point));
+    } else if (info.isDir() && index.data().toString() != ".." && !m_s3model->getRootBucket().isEmpty()) {
+        menu.addAction("Upload", this, SLOT(on_uploadDir()));
+        menu.exec(m_fsview->mapToGlobal(point));
     } else {
         return;
     }
 }
 
+QFileInfoList GetFileList(QString path)
```
Comment (thesues) [inline src/S3ClientGUI/mainwindow.cpp:368]: check this: https://stackoverflow.com/questions/8052460/recursively-iterate-over-all-the-files-in-a-directory-and-its-subdirectories-in

this iterator way is much much better

File: src/S3ClientGUI/mainwindow.cpp [hunk 3]
```diff
@@ -356,12 +357,83 @@ void MainWindow::on_LocalContextMenuRequest(const QPoint &point)
...
+{
+    QFileInfoList fil;
+    QFileInfo info = m_fsview->currentFileInfo();
+    QString parentDirPath = info.absolutePath() + '/';
+    qDebug() << "info=" << qPrintable(info.absolutePath()) << endl;
+    QString remoteBucketName = m_s3model->getRootBucket();
+    QString remotePrefix = m_s3model->getCurrentPrefix();
+    QString KeyName;
+
+    fil = GetFileList(info.absoluteFilePath());
```
Comment (thesues) [inline src/S3ClientGUI/mainwindow.cpp:396]: what is ```fil```?

File: src/S3ClientGUI/mainwindow.cpp [hunk 4]
```diff
@@ -356,12 +357,83 @@ void MainWindow::on_LocalContextMenuRequest(const QPoint &point)
...
+    for (int i = 0; i < fil.size(); i++) {
+        qDebug() << "fil=" << fil.at(i) << endl;
+    }
+    for (int i = 0; i < fil.size(); ++i) {
```
Comment (thesues) [inline src/S3ClientGUI/mainwindow.cpp:400]: use iterator.., something like for(auto &file: list) {}

File: src/S3ClientGUI/mainwindow.cpp [hunk 5]
```diff
@@ -356,12 +357,83 @@ void MainWindow::on_LocalContextMenuRequest(const QPoint &point)
...
+        QString localFile = fil.at(i).absoluteFilePath();
+        QString tmp = localFile;
+        tmp.remove(0, parentDirPath.length() - 1);
+        KeyName = remotePrefix + tmp;
+        qDebug() << "tmp=" << qPrintable(tmp) << endl;
+        qDebug() << "KeyName=" << qPrintable(KeyName) << endl;
+        //if dir is empty,mkdir!
+        if (!fil.at(i).isFile()) {
```
Comment (thesues) [inline src/S3ClientGUI/mainwindow.cpp:408]: What? why create the directory for S3?

File: src/S3ClientGUI/mainwindow.h [hunk 6]
```diff
@@ -65,6 +67,11 @@ private slots:
     QProgressDialog *quitDialog;
     QMenu *fileMenu;
     QAction *exitAction;
+
+    QStringList file_list;
```
Comment (thesues) [inline src/S3ClientGUI/mainwindow.h:71]: fuck, file_list can not be global

File: src/S3ClientGUI/mainwindow.cpp [hunk 7]
```diff
@@ -485,16 +557,94 @@ void MainWindow::on_download()
...
+            qDebug() << "parent Dir=" << tmp << endl;
+            QDir dir(tmp);
+            if (!dir.exists()) {
+                dir.mkpath(tmp);
+            }
+            if (downloadFile.endsWith("/"))
+                continue;
+            qDebug() << "tmp=" << tmp << endl;
+            qDebug() << "downloadFile=" << downloadFile << endl;
+            DownloadObjectHandler *pHandler = m_s3client->DownloadFile(bucketName, file_list.at(i),
```
Comment (thesues) [inline src/S3ClientGUI/mainwindow.cpp:628]: Can you put  this "Create a new download/upload snippet" into a new function?

File: src/S3ClientGUI/qtaskmodel.cpp [hunk 8]
```diff
@@ -103,6 +111,25 @@ Qt::ItemFlags QTaskModel::flags(const QModelIndex &index) const
     return (QAbstractItemModel::flags(index)) & ~ Qt::ItemIsEditable;
 }
 
+void QTaskModel::reDownload(QModelIndex &index)
```
Comment (thesues) [inline src/S3ClientGUI/qtaskmodel.cpp:114]: No, you should create a "New task" , I do not like the idea of reusing the same s3 instance

File: src/S3ClientGUI/mainwindow.cpp (hunk 5 above)
Comment (Zstro) [reply src/S3ClientGUI/mainwindow.cpp:408]: Maybe someone directory has empty directory.Should I delete it?

**Pull request closed without merging**